xonsh/completers/tools.py

```diff
@@ -31,3 +31,53 @@ def justify(s, max_length, left_pad=0):
     """
     txt = textwrap.wrap(s, width=max_length, subsequent_indent=" " * left_pad)
     return "\n".join(txt)
+
+
+class RichCompletion(str):
+    """A rich completion that completers can return instead of a string
+
+    Parameters
+    ----------
+    value: str
+        The completion's actual value.
+    prefix_len : int
+        Length of the prefix to be replaced in the completion.
+        If None, the default prefix len will be used.
+    display: str
+        Text to display in completion option list.
+        If None, `value` will be used.
+    description : str
+        Extra text to display when the completion is selected.
+    """
+
+    def __new__(cls, value, prefix_len=None, display=None, description=""):
+        completion = super().__new__(cls, value)
+
+        completion.prefix_len = prefix_len
+        completion.display = display or value
+        completion.description = description
+
+        return completion
+
+
+def get_ptk_completer():
+    """Get the current PromptToolkitCompleter
+
+    This is usefull for completers that want to use
+    PromptToolkitCompleter.current_document (the current multiline document).
+
+    Call this function lazily since in '.xonshrc' the shell doesn't exist.
+
+    Returns
+    -------
+    The PromptToolkitCompleter if running with ptk, else returns None
+    """
+    if __xonsh__.shell is None:
+        return None
+
+    shell = __xonsh__.shell.shell
+
+    # avoid importing the class in case ptk isn't installed
+    if shell.__class__.__name__ == "PromptToolkitShell":
```

Comment:
I think it is more robust to check `__xonsh__.shell.shell_type` than the class name

Reply:
done